apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsLinkControl.tsx

```diff
@@ -0,0 +1,117 @@
+import type { ControlProps, RankedTester } from "@jsonforms/core"
+import {
+  Box,
+  FormControl,
+  HStack,
+  Icon,
+  TabList,
+  TabPanel,
+  TabPanels,
+  Tabs,
+  Text,
+} from "@chakra-ui/react"
+import { and, isStringControl, rankWith, schemaMatches } from "@jsonforms/core"
+import { withJsonFormsControlProps } from "@jsonforms/react"
+import { FormLabel, Input, Tab } from "@opengovsg/design-system-react"
+import { BiEnvelopeOpen, BiFile, BiFileBlank, BiLink } from "react-icons/bi"
+
+import { JSON_FORMS_RANKING } from "~/constants/formBuilder"
+
+export const jsonFormsLinkControlTester: RankedTester = rankWith(
+  JSON_FORMS_RANKING.LinkControl,
+  and(
+    isStringControl,
+    schemaMatches((schema) => schema.format === "link"),
+  ),
+)
+
+export function JsonFormsLinkControl({
+  data,
+  label,
+  handleChange,
+  path,
+  description,
+  required,
+}: ControlProps) {
+  const LINK_TYPES = [
+    {
+      icon: BiFileBlank,
+      label: "Page",
+    },
+    {
+      icon: BiLink,
+      label: "External link",
+    },
+    {
+      icon: BiFile,
+      label: "File",
+    },
+    {
+      icon: BiEnvelopeOpen,
+      label: "Email",
+    },
+  ]
+  const dataString = data && typeof data === "string" ? data : ""
+
+  return (
+    <Box py={2}>
+      <FormControl isRequired={required}>
+        <FormLabel description={description}>{label}</FormLabel>
+
+        <Tabs variant="group" onChange={() => handleChange(path, "")}>
+          <TabList>
+            {LINK_TYPES.map(({ icon, label }) => (
+              <Tab key={label}>
+                <HStack spacing={2}>
+                  <Icon as={icon} fontSize="1.25rem" />
+                  <Text>{label}</Text>
+                </HStack>
+              </Tab>
+            ))}
+          </TabList>
+
+          <TabPanels>
+            <TabPanel>
+              <Input
+                type="text"
+                value={dataString}
+                onChange={(e) => handleChange(path, e.target.value)}
+                placeholder="Page permalink"
+              />
```

Comment:
this however is the bigger issue (and thus i'd recommend dynamic state based on selected radio value instead), as the tab panels can desynchronise with `LINK_TYPES` and a specific order is required, which is not obvious if `LINK_TYPE` is ever modified.

might be safer to not use tabs and just render a specific input depending on the selected radio option. to that end, you can rely on compile time typechecking to ensure something is rendered accordingly to the option.

UNLESS the value of the current tab does not matter to you, and only the input value does. then maybe tab can be used.

Reply:
Updated in 950c8ed0b863961599692a2057cc26ee605d469c, agreed that it would better with a more explicit mapping of the link type to the actual rendering.

The value of the current tab would not matter because the 4 link types are mutually exclusive, so it is not possible to translate from one to another. Opted to clear out the value of the input to avoid confusing the user.